runtime/gc_base/StringTable.cpp

```diff
@@ -563,7 +575,16 @@ j9gc_createJavaLangString(J9VMThread *vmThread, U_8 *data, UDATA length, UDATA s
 	 */
 
 	if ((stringFlags & (J9_STR_XLAT | J9_STR_UNICODE)) == 0) {
-		U_32 hash = (U_32)vm->internalVMFunctions->computeHashForUTF8(data, length);
+		UDATA hash = 0;
+
+		if (isCompressible) {
+			for (UDATA i = 0; i < length; ++i) {
+				hash = (hash << 5) - hash + data[i];
+			}
+		} else {
+			hash = (U_32)vm->internalVMFunctions->computeHashForUTF8(data, length);
+		}
```

Comment:
A better answer here is to add an implementation of computeHashForUTF8 to VMHelpers.hpp and call that from here.  The compiler will do the inlining and we avoid spreading the implementation all over.

Reply:
Unfortunately at least on Linux on Z GCC is unable to inline past the function pointer table. Is there some place we can put these APIs so that we don't have to indirect via a function table? I would also like to change the existing API to be inlined as well.

As proof we can see that the existing API is not inlined in the disassembly:
```
root@bcf5e9a8d8d7:~/openj9-openjdk-jdk11# objdump -d ./build/linux-s390x-normal-server-release/vm/gc_base/StringTable.o | sed '/<j9gc_createJavaLangString>:/,/^$/!d'
00000000000028f0 <j9gc_createJavaLangString>:
    28f0:       eb 6f f0 30 00 24       stmg    %r6,%r15,48(%r15)
    28f6:       c0 d0 00 00 00 00       larl    %r13,28f6 <j9gc_createJavaLangString+0x6>
    28fc:       e3 f0 fe f8 ff 71       lay     %r15,-264(%r15)
    2902:       e3 70 20 08 00 04       lg      %r7,8(%r2)
    2908:       92 01 f1 07             mvi     263(%r15),1
    290c:       e3 50 f0 b0 00 24       stg     %r5,176(%r15)
    2912:       b9 04 00 b2             lgr     %r11,%r2
    2916:       b9 04 00 a3             lgr     %r10,%r3
    291a:       e3 10 7d 10 04 04       lg      %r1,19728(%r7)
    2920:       e3 10 10 60 00 04       lg      %r1,96(%r1)
    2926:       e3 10 10 f0 06 04       lg      %r1,24816(%r1)
    292c:       e3 10 f0 d8 00 24       stg     %r1,216(%r15)
    2932:       b9 02 00 14             ltgr    %r1,%r4
    2936:       e3 10 f0 c0 00 24       stg     %r1,192(%r15)
    293c:       a7 84 00 0f             je      295a <j9gc_createJavaLangString+0x6a>
    2940:       91 80 30 00             tm      0(%r3),128
    2944:       a7 74 00 dd             jne     2afe <j9gc_createJavaLangString+0x20e>
    2948:       b9 04 00 45             lgr     %r4,%r5
    294c:       41 20 30 01             la      %r2,1(%r3)
    2950:       a7 17 00 cb             brctg   %r1,2ae6 <j9gc_createJavaLangString+0x1f6>
    2954:       e3 40 f0 b0 00 24       stg     %r4,176(%r15)
    295a:       96 80 f0 b7             oi      183(%r15),128
    295e:       c0 10 00 00 00 00       larl    %r1,295e <j9gc_createJavaLangString+0x6e>
    2964:       e3 10 f0 e8 00 24       stg     %r1,232(%r15)
    296a:       e3 10 10 00 00 04       lg      %r1,0(%r1)
    2970:       e3 40 11 61 00 94       llc     %r4,353(%r1)
    2976:       12 44                   ltr     %r4,%r4
    2978:       a7 84 00 27             je      29c6 <j9gc_createJavaLangString+0xd6>
    297c:       c0 10 00 00 00 00       larl    %r1,297c <j9gc_createJavaLangString+0x8c>
    2982:       c0 4d 00 01 61 00       oilf    %r4,90368
    2988:       e3 00 f0 b0 00 04       lg      %r0,176(%r15)
    298e:       e3 60 f0 c0 00 04       lg      %r6,192(%r15)
    2994:       c0 50 00 00 00 00       larl    %r5,2994 <j9gc_createJavaLangString+0xa4>
    299a:       b9 04 00 2b             lgr     %r2,%r11
    299e:       e3 10 10 00 00 04       lg      %r1,0(%r1)
    29a4:       b9 04 00 31             lgr     %r3,%r1
    29a8:       b9 16 00 44             llgfr   %r4,%r4
    29ac:       e3 10 10 20 00 04       lg      %r1,32(%r1)
    29b2:       e3 00 f0 a8 00 24       stg     %r0,168(%r15)
    29b8:       e3 a0 f0 a0 00 24       stg     %r10,160(%r15)
    29be:       e3 10 10 00 00 04       lg      %r1,0(%r1)
    29c4:       0d e1                   basr    %r14,%r1
    29c6:       a7 19 00 11             lghi    %r1,17
    29ca:       e3 10 f0 b0 00 80       ng      %r1,176(%r15)
    29d0:       a7 74 00 9b             jne     2b06 <j9gc_createJavaLangString+0x216>
    29d4:       95 00 f1 07             cli     263(%r15),0
    29d8:       a7 74 02 46             jne     2e64 <j9gc_createJavaLangString+0x574>
    29dc:       e3 10 70 00 00 04       lg      %r1,0(%r7)
    29e2:       e3 30 f0 c0 00 04       lg      %r3,192(%r15)
    29e8:       b9 04 00 2a             lgr     %r2,%r10
    29ec:       e3 10 11 18 00 04       lg      %r1,280(%r1)
    29f2:       0d e1                   basr    %r14,%r1
    29f4:       b9 04 00 12             lgr     %r1,%r2
    29f8:       c0 1a 00 00 00 00       nihf    %r1,0
    29fe:       e3 10 f0 d0 00 24       stg     %r1,208(%r15)
    2a04:       b9 16 00 62             llgfr   %r6,%r2
    2a08:       a7 29 00 00             lghi    %r2,0
    2a0c:       e3 20 f0 c8 00 24       stg     %r2,200(%r15)
    2a12:       e3 80 f0 d8 00 04       lg      %r8,216(%r15)
    2a18:       eb 23 f0 c8 00 04       lmg     %r2,%r3,200(%r15)
    2a1e:       e3 10 80 20 00 04       lg      %r1,32(%r8)
    2a24:       e3 20 80 10 00 87       dlg     %r2,16(%r8)
    2a2a:       eb 23 f0 c8 00 24       stmg    %r2,%r3,200(%r15)
    2a30:       eb 92 00 03 00 0d       sllg    %r9,%r2,3
    2a36:       e3 29 10 00 00 04       lg      %r2,0(%r9,%r1)
    2a3c:       c0 e5 00 00 00 00       brasl   %r14,2a3c <j9gc_createJavaLangString+0x14c>
    2a42:       e3 80 f0 d8 00 24       stg     %r8,216(%r15)
    2a48:       e3 30 f0 c8 00 04       lg      %r3,200(%r15)
    2a4e:       e3 50 f0 c0 00 04       lg      %r5,192(%r15)
    2a54:       b9 04 00 4a             lgr     %r4,%r10
    2a58:       b9 04 00 28             lgr     %r2,%r8
    2a5c:       c0 e5 00 00 00 00       brasl   %r14,2a5c <j9gc_createJavaLangString+0x16c>
    2a62:       e3 10 f0 d8 00 04       lg      %r1,216(%r15)
    2a68:       e3 10 10 20 00 04       lg      %r1,32(%r1)
    2a6e:       b9 04 00 82             lgr     %r8,%r2
    2a72:       e3 29 10 00 00 04       lg      %r2,0(%r9,%r1)
    2a78:       c0 e5 00 00 00 00       brasl   %r14,2a78 <j9gc_createJavaLangString+0x188>
    2a7e:       b9 02 00 88             ltgr    %r8,%r8
    2a82:       a7 84 00 42             je      2b06 <j9gc_createJavaLangString+0x216>
    2a86:       e3 10 f0 e8 00 04       lg      %r1,232(%r15)
    2a8c:       e3 10 10 00 00 04       lg      %r1,0(%r1)
    2a92:       e3 40 11 62 00 94       llc     %r4,354(%r1)
    2a98:       12 44                   ltr     %r4,%r4
    2a9a:       a7 84 00 1d             je      2ad4 <j9gc_createJavaLangString+0x1e4>
    2a9e:       c0 10 00 00 00 00       larl    %r1,2a9e <j9gc_createJavaLangString+0x1ae>
    2aa4:       c0 4d 00 01 62 00       oilf    %r4,90624
    2aaa:       b9 04 00 2b             lgr     %r2,%r11
    2aae:       b9 04 00 68             lgr     %r6,%r8
    2ab2:       c0 50 00 00 00 00       larl    %r5,2ab2 <j9gc_createJavaLangString+0x1c2>
    2ab8:       e3 10 10 00 00 04       lg      %r1,0(%r1)
    2abe:       b9 04 00 31             lgr     %r3,%r1
    2ac2:       b9 16 00 44             llgfr   %r4,%r4
    2ac6:       e3 10 10 20 00 04       lg      %r1,32(%r1)
    2acc:       e3 10 10 00 00 04       lg      %r1,0(%r1)
    2ad2:       0d e1                   basr    %r14,%r1
    2ad4:       e3 40 f1 78 00 04       lg      %r4,376(%r15)
    2ada:       b9 04 00 28             lgr     %r2,%r8
    2ade:       eb 6f f1 38 00 04       lmg     %r6,%r15,312(%r15)
    2ae4:       07 f4                   br      %r4
    2ae6:       e3 30 20 00 00 90       llgc    %r3,0(%r2)
    2aec:       a7 2b 00 01             aghi    %r2,1
    2af0:       a7 31 00 80             tmll    %r3,128
    2af4:       a7 84 ff 2e             je      2950 <j9gc_createJavaLangString+0x60>
    2af8:       e3 40 f0 b0 00 24       stg     %r4,176(%r15)
    2afe:       92 00 f1 07             mvi     263(%r15),0
    2b02:       a7 f4 ff 2e             j       295e <j9gc_createJavaLangString+0x6e>
    2b06:       e3 20 f0 b0 00 04       lg      %r2,176(%r15)
    2b0c:       eb 12 00 02 00 0c       srlg    %r1,%r2,2
    2b12:       a7 39 00 01             lghi    %r3,1
    2b16:       b9 80 00 13             ngr     %r1,%r3
    2b1a:       e3 10 f0 e0 00 24       stg     %r1,224(%r15)
    2b20:       a7 21 00 0a             tmll    %r2,10
    2b24:       a7 84 00 04             je      2b2c <j9gc_createJavaLangString+0x23c>
    2b28:       96 02 f0 e7             oi      231(%r15),2
    2b2c:       e3 10 70 00 00 04       lg      %r1,0(%r7)
    2b32:       a7 49 00 02             lghi    %r4,2
    2b36:       a7 39 00 18             lghi    %r3,24
    2b3a:       b9 04 00 2b             lgr     %r2,%r11
    2b3e:       e3 10 11 08 00 04       lg      %r1,264(%r1)
    2b44:       0d e1                   basr    %r14,%r1
    2b46:       b9 02 00 12             ltgr    %r1,%r2
    2b4a:       e3 10 f0 f8 00 24       stg     %r1,248(%r15)
    2b50:       a7 84 01 5b             je      2e06 <j9gc_createJavaLangString+0x516>
    2b54:       e3 40 f0 e0 00 04       lg      %r4,224(%r15)
    2b5a:       b9 04 00 31             lgr     %r3,%r1
    2b5e:       b9 04 00 2b             lgr     %r2,%r11
    2b62:       c0 e5 00 00 00 00       brasl   %r14,2b62 <j9gc_createJavaLangString+0x272>
    2b68:       b9 02 00 92             ltgr    %r9,%r2
    2b6c:       a7 84 01 4d             je      2e06 <j9gc_createJavaLangString+0x516>
    2b70:       a7 69 00 10             lghi    %r6,16
    2b74:       e3 30 f0 b0 00 04       lg      %r3,176(%r15)
    2b7a:       b9 80 00 63             ngr     %r6,%r3
    2b7e:       e3 10 7c b8 04 02       ltg     %r1,19640(%r7)
    2b84:       a7 84 00 34             je      2bec <j9gc_createJavaLangString+0x2fc>
    2b88:       b9 02 00 66             ltgr    %r6,%r6
    2b8c:       a7 84 01 8f             je      2eaa <j9gc_createJavaLangString+0x5ba>
    2b90:       e3 10 f0 c0 00 04       lg      %r1,192(%r15)
    2b96:       eb 11 00 01 00 0c       srlg    %r1,%r1,1
    2b9c:       e3 10 f0 c8 00 24       stg     %r1,200(%r15)
    2ba2:       b9 02 00 11             ltgr    %r1,%r1
    2ba6:       a7 84 04 37             je      3414 <j9gc_createJavaLangString+0xb24>
    2baa:       d5 01 a0 00 d0 02       clc     0(2,%r10),2(%r13)
    2bb0:       a7 24 01 49             jh      2e42 <j9gc_createJavaLangString+0x552>
    2bb4:       eb 11 00 01 00 0d       sllg    %r1,%r1,1
    2bba:       41 20 a0 02             la      %r2,2(%r10)
    2bbe:       b9 04 00 43             lgr     %r4,%r3
    2bc2:       e3 10 1f fe ff 71       lay     %r1,-2(%r1)
    2bc8:       eb 11 00 01 00 0c       srlg    %r1,%r1,1
    2bce:       a7 1b 00 01             aghi    %r1,1
    2bd2:       a7 2b 00 02             aghi    %r2,2
    2bd6:       a7 17 01 2c             brctg   %r1,2e2e <j9gc_createJavaLangString+0x53e>
    2bda:       e3 40 f0 b0 00 24       stg     %r4,176(%r15)
    2be0:       a7 18 00 01             lhi     %r1,1
    2be4:       42 10 f1 07             stc     %r1,263(%r15)
    2be8:       a7 f4 00 10             j       2c08 <j9gc_createJavaLangString+0x318>
    2bec:       b9 02 00 66             ltgr    %r6,%r6
    2bf0:       a7 74 01 2f             jne     2e4e <j9gc_createJavaLangString+0x55e>
    2bf4:       95 00 f1 07             cli     263(%r15),0
    2bf8:       a7 84 03 49             je      328a <j9gc_createJavaLangString+0x99a>
    2bfc:       e3 10 f0 c0 00 04       lg      %r1,192(%r15)
    2c02:       e3 10 f0 c8 00 24       stg     %r1,200(%r15)
    2c08:       e3 10 b0 20 00 04       lg      %r1,32(%r11)
    2c0e:       e3 20 1f f8 ff 71       lay     %r2,-8(%r1)
    2c14:       e3 20 b0 20 00 24       stg     %r2,32(%r11)
    2c1a:       e3 90 1f f8 ff 24       stg     %r9,-8(%r1)
    2c20:       58 20 f0 cc             l       %r2,204(%r15)
    2c24:       50 20 f0 f4             st      %r2,244(%r15)
    2c28:       e3 10 b0 30 00 04       lg      %r1,48(%r11)
    2c2e:       a7 1b 00 08             aghi    %r1,8
    2c32:       e3 10 b0 30 00 24       stg     %r1,48(%r11)
    2c38:       91 80 70 cb             tm      203(%r7),128
    2c3c:       a7 84 01 a5             je      2f86 <j9gc_createJavaLangString+0x696>
    2c40:       e3 10 7c b8 04 02       ltg     %r1,19640(%r7)
    2c46:       e3 30 72 10 00 04       lg      %r3,528(%r7)
    2c4c:       e3 50 f0 e0 00 04       lg      %r5,224(%r15)
    2c52:       a7 84 00 06             je      2c5e <j9gc_createJavaLangString+0x36e>
    2c56:       95 00 f1 07             cli     263(%r15),0
    2c5a:       a7 74 03 0d             jne     3274 <j9gc_createJavaLangString+0x984>
    2c5e:       58 40 f0 f4             l       %r4,244(%r15)
    2c62:       89 40 00 01             sll     %r4,1
    2c66:       b9 04 00 2b             lgr     %r2,%r11
    2c6a:       b9 16 00 44             llgfr   %r4,%r4
    2c6e:       c0 e5 00 00 00 00       brasl   %r14,2c6e <j9gc_createJavaLangString+0x37e>
    2c74:       b9 04 00 92             lgr     %r9,%r2
    2c78:       e3 10 b0 30 00 04       lg      %r1,48(%r11)
    2c7e:       a7 1b ff f8             aghi    %r1,-8
    2c82:       e3 10 b0 30 00 24       stg     %r1,48(%r11)
    2c88:       e3 10 b0 20 00 04       lg      %r1,32(%r11)
    2c8e:       b9 02 00 99             ltgr    %r9,%r9
    2c92:       41 20 10 08             la      %r2,8(%r1)
    2c96:       e3 20 b0 20 00 24       stg     %r2,32(%r11)
    2c9c:       e3 80 10 00 00 04       lg      %r8,0(%r1)
    2ca2:       a7 84 00 b2             je      2e06 <j9gc_createJavaLangString+0x516>
    2ca6:       b9 02 00 66             ltgr    %r6,%r6
    2caa:       e3 10 7c b8 04 04       lg      %r1,19640(%r7)
    2cb0:       a7 84 01 86             je      2fbc <j9gc_createJavaLangString+0x6cc>
    2cb4:       b9 02 00 11             ltgr    %r1,%r1
    2cb8:       e3 10 f0 c8 00 04       lg      %r1,200(%r15)
    2cbe:       a7 84 01 05             je      2ec8 <j9gc_createJavaLangString+0x5d8>
    2cc2:       95 00 f1 07             cli     263(%r15),0
    2cc6:       a7 84 01 01             je      2ec8 <j9gc_createJavaLangString+0x5d8>
    2cca:       b9 02 00 11             ltgr    %r1,%r1
    2cce:       a7 84 00 26             je      2d1a <j9gc_createJavaLangString+0x42a>
    2cd2:       a7 19 00 00             lghi    %r1,0
    2cd6:       e3 00 f0 b0 00 04       lg      %r0,176(%r15)
    2cdc:       b9 16 00 21             llgfr   %r2,%r1
    2ce0:       e3 20 f0 b8 00 24       stg     %r2,184(%r15)
    2ce6:       e3 20 90 04 00 12       lt      %r2,4(%r9)
    2cec:       48 40 a0 00             lh      %r4,0(%r10)
    2cf0:       a7 29 00 00             lghi    %r2,0
    2cf4:       e3 20 f0 b0 00 24       stg     %r2,176(%r15)
    2cfa:       a7 84 02 80             je      31fa <j9gc_createJavaLangString+0x90a>
    2cfe:       42 41 90 08             stc     %r4,8(%r1,%r9)
    2d02:       a7 1b 00 01             aghi    %r1,1
    2d06:       a7 ab 00 02             aghi    %r10,2
    2d0a:       e3 10 f0 c8 00 20       cg      %r1,200(%r15)
    2d10:       a7 74 ff e6             jne     2cdc <j9gc_createJavaLangString+0x3ec>
    2d14:       e3 00 f0 b0 00 24       stg     %r0,176(%r15)
    2d1a:       e3 10 b0 08 00 04       lg      %r1,8(%r11)
    2d20:       e3 20 13 18 02 04       lg      %r2,8984(%r1)
    2d26:       a7 2b ff f9             aghi    %r2,-7
    2d2a:       c2 2e 00 00 00 01       clgfi   %r2,1
    2d30:       a7 24 00 17             jh      2d5e <j9gc_createJavaLangString+0x46e>
    2d34:       e3 40 19 f0 00 04       lg      %r4,2544(%r1)
    2d3a:       e3 10 11 30 00 04       lg      %r1,304(%r1)
    2d40:       b9 04 00 59             lgr     %r5,%r9
    2d44:       b9 04 00 38             lgr     %r3,%r8
    2d48:       b9 04 00 2b             lgr     %r2,%r11
    2d4c:       41 44 80 04             la      %r4,4(%r4,%r8)
    2d50:       e3 10 11 f0 00 04       lg      %r1,496(%r1)
    2d56:       0d e1                   basr    %r14,%r1
    2d58:       e3 10 b0 08 00 04       lg      %r1,8(%r11)
    2d5e:       e3 20 19 f0 00 04       lg      %r2,2544(%r1)
    2d64:       e3 10 1b 40 04 04       lg      %r1,19264(%r1)
    2d6a:       eb 19 10 00 00 0c       srlg    %r1,%r9,0(%r1)
    2d70:       50 12 80 04             st      %r1,4(%r2,%r8)
    2d74:       e3 10 b0 08 00 04       lg      %r1,8(%r11)
    2d7a:       e3 20 13 10 02 04       lg      %r2,8976(%r1)
    2d80:       a7 2f 00 02             cghi    %r2,2
    2d84:       a7 84 00 12             je      2da8 <j9gc_createJavaLangString+0x4b8>
    2d88:       e3 10 11 30 00 04       lg      %r1,304(%r1)
    2d8e:       b9 04 00 49             lgr     %r4,%r9
    2d92:       b9 04 00 38             lgr     %r3,%r8
    2d96:       b9 04 00 2b             lgr     %r2,%r11
    2d9a:       e3 10 10 20 00 04       lg      %r1,32(%r1)
    2da0:       0d e1                   basr    %r14,%r1
    2da2:       e3 10 b0 08 00 04       lg      %r1,8(%r11)
    2da8:       a5 2f ff ff             llill   %r2,65535
    2dac:       e3 20 70 28 00 80       ng      %r2,40(%r7)
    2db2:       e3 30 7c b8 04 02       ltg     %r3,19640(%r7)
    2db8:       a7 84 01 27             je      3006 <j9gc_createJavaLangString+0x716>
    2dbc:       95 00 f1 07             cli     263(%r15),0
    2dc0:       a7 84 01 31             je      3022 <j9gc_createJavaLangString+0x732>
    2dc4:       c2 2e 00 00 0a ff       clgfi   %r2,2815
    2dca:       a7 24 02 0f             jh      31e8 <j9gc_createJavaLangString+0x8f8>
    2dce:       e3 10 1a 00 00 04       lg      %r1,2560(%r1)
    2dd4:       58 20 f0 f4             l       %r2,244(%r15)
    2dd8:       50 21 80 04             st      %r2,4(%r1,%r8)
    2ddc:       e3 10 f0 b0 00 04       lg      %r1,176(%r15)
    2de2:       a7 11 00 08             tmll    %r1,8
    2de6:       a7 84 fe 50             je      2a86 <j9gc_createJavaLangString+0x196>
    2dea:       b9 04 00 48             lgr     %r4,%r8
    2dee:       e3 20 f0 d8 00 04       lg      %r2,216(%r15)
    2df4:       b9 04 00 3b             lgr     %r3,%r11
    2df8:       c0 e5 00 00 00 00       brasl   %r14,2df8 <j9gc_createJavaLangString+0x508>
    2dfe:       b9 02 00 82             ltgr    %r8,%r2
    2e02:       a7 74 fe 42             jne     2a86 <j9gc_createJavaLangString+0x196>
    2e06:       e3 10 70 00 00 04       lg      %r1,0(%r7)
    2e0c:       b9 04 00 2b             lgr     %r2,%r11
    2e10:       a7 89 00 00             lghi    %r8,0
    2e14:       e3 10 17 d8 00 04       lg      %r1,2008(%r1)
    2e1a:       0d e1                   basr    %r14,%r1
    2e1c:       e3 40 f1 78 00 04       lg      %r4,376(%r15)
    2e22:       b9 04 00 28             lgr     %r2,%r8
    2e26:       eb 6f f1 38 00 04       lmg     %r6,%r15,312(%r15)
    2e2c:       07 f4                   br      %r4
    2e2e:       e3 30 2f fe ff 78       lhy     %r3,-2(%r2)
    2e34:       bd 33 d0 02             clm     %r3,3,2(%r13)
    2e38:       a7 c4 fe cd             jle     2bd2 <j9gc_createJavaLangString+0x2e2>
    2e3c:       e3 40 f0 b0 00 24       stg     %r4,176(%r15)
    2e42:       a7 18 00 00             lhi     %r1,0
    2e46:       42 10 f1 07             stc     %r1,263(%r15)
    2e4a:       a7 f4 fe df             j       2c08 <j9gc_createJavaLangString+0x318>
    2e4e:       e3 10 f0 c0 00 04       lg      %r1,192(%r15)
    2e54:       eb 11 00 01 00 0c       srlg    %r1,%r1,1
    2e5a:       e3 10 f0 c8 00 24       stg     %r1,200(%r15)
    2e60:       a7 f4 fe d4             j       2c08 <j9gc_createJavaLangString+0x318>
    2e64:       e3 30 f0 c0 00 02       ltg     %r3,192(%r15)
    2e6a:       a7 84 02 cf             je      3408 <j9gc_createJavaLangString+0xb18>
    2e6e:       b9 04 00 4a             lgr     %r4,%r10
    2e72:       b9 04 00 21             lgr     %r2,%r1
    2e76:       e3 00 f0 b0 00 04       lg      %r0,176(%r15)
    2e7c:       eb 12 00 05 00 0d       sllg    %r1,%r2,5
    2e82:       e3 50 40 00 00 90       llgc    %r5,0(%r4)
    2e88:       a7 4b 00 01             aghi    %r4,1
    2e8c:       b9 09 00 12             sgr     %r1,%r2
    2e90:       b9 04 00 25             lgr     %r2,%r5
    2e94:       b9 08 00 21             agr     %r2,%r1
    2e98:       a7 37 ff f2             brctg   %r3,2e7c <j9gc_createJavaLangString+0x58c>
    2e9c:       e3 00 f0 b0 00 24       stg     %r0,176(%r15)
    2ea2:       b9 04 00 12             lgr     %r1,%r2
    2ea6:       a7 f4 fd ac             j       29fe <j9gc_createJavaLangString+0x10e>
    2eaa:       41 40 f1 07             la      %r4,263(%r15)
    2eae:       e3 30 f0 c0 00 04       lg      %r3,192(%r15)
    2eb4:       b9 04 00 2a             lgr     %r2,%r10
    2eb8:       c0 e5 ff ff e8 a4       brasl   %r14,0 <getUnicodeLength>
    2ebe:       e3 20 f0 c8 00 24       stg     %r2,200(%r15)
    2ec4:       a7 f4 fe a2             j       2c08 <j9gc_createJavaLangString+0x318>
    2ec8:       b9 02 00 11             ltgr    %r1,%r1
    2ecc:       a7 84 ff 27             je      2d1a <j9gc_createJavaLangString+0x42a>
    2ed0:       a7 19 00 00             lghi    %r1,0
    2ed4:       e3 00 f0 b0 00 04       lg      %r0,176(%r15)
    2eda:       b9 16 00 21             llgfr   %r2,%r1
    2ede:       e3 20 f0 b8 00 24       stg     %r2,184(%r15)
    2ee4:       e3 20 90 04 00 12       lt      %r2,4(%r9)
    2eea:       eb 41 00 01 00 0d       sllg    %r4,%r1,1
    2ef0:       a7 29 00 00             lghi    %r2,0
    2ef4:       e3 20 f0 b0 00 24       stg     %r2,176(%r15)
    2efa:       a7 84 00 11             je      2f1c <j9gc_createJavaLangString+0x62c>
    2efe:       a7 1b 00 01             aghi    %r1,1
    2f02:       48 20 a0 00             lh      %r2,0(%r10)
    2f06:       a7 ab 00 02             aghi    %r10,2
    2f0a:       40 24 90 08             sth     %r2,8(%r4,%r9)
    2f0e:       e3 10 f0 c8 00 20       cg      %r1,200(%r15)
    2f14:       a7 74 ff e3             jne     2eda <j9gc_createJavaLangString+0x5ea>
    2f18:       a7 f4 fe fe             j       2d14 <j9gc_createJavaLangString+0x424>
    2f1c:       e3 50 b0 08 00 04       lg      %r5,8(%r11)
    2f22:       a7 1b 00 01             aghi    %r1,1
    2f26:       eb 23 f0 b0 00 04       lmg     %r2,%r3,176(%r15)
    2f2c:       a7 ab 00 02             aghi    %r10,2
    2f30:       e3 40 5b e8 04 04       lg      %r4,19432(%r5)
    2f36:       eb 44 00 01 00 0c       srlg    %r4,%r4,1
    2f3c:       e3 50 5b 40 04 04       lg      %r5,19264(%r5)
    2f42:       b9 87 00 24             dlgr    %r2,%r4
    2f46:       e3 40 af fe ff 78       lhy     %r4,-2(%r10)
    2f4c:       eb 23 f0 b0 00 24       stmg    %r2,%r3,176(%r15)
    2f52:       a7 3b 00 04             aghi    %r3,4
    2f56:       e3 10 f0 c8 00 20       cg      %r1,200(%r15)
    2f5c:       eb 33 00 02 00 0d       sllg    %r3,%r3,2
    2f62:       e3 20 f0 b0 00 04       lg      %r2,176(%r15)
    2f68:       eb 22 00 01 00 0d       sllg    %r2,%r2,1
    2f6e:       e3 33 90 00 00 16       llgf    %r3,0(%r3,%r9)
    2f74:       eb 33 50 00 00 0d       sllg    %r3,%r3,0(%r5)
    2f7a:       40 42 30 00             sth     %r4,0(%r2,%r3)
    2f7e:       a7 74 ff ae             jne     2eda <j9gc_createJavaLangString+0x5ea>
    2f82:       a7 f4 fe c9             j       2d14 <j9gc_createJavaLangString+0x424>
    2f86:       e3 10 7c b8 04 02       ltg     %r1,19640(%r7)
    2f8c:       e3 30 71 f8 00 04       lg      %r3,504(%r7)
    2f92:       a7 84 00 06             je      2f9e <j9gc_createJavaLangString+0x6ae>
    2f96:       95 00 f1 07             cli     263(%r15),0
    2f9a:       a7 74 01 5a             jne     324e <j9gc_createJavaLangString+0x95e>
    2f9e:       e3 50 f0 e0 00 04       lg      %r5,224(%r15)
    2fa4:       e3 40 f0 f4 00 16       llgf    %r4,244(%r15)
    2faa:       b9 04 00 2b             lgr     %r2,%r11
    2fae:       c0 e5 00 00 00 00       brasl   %r14,2fae <j9gc_createJavaLangString+0x6be>
    2fb4:       b9 04 00 92             lgr     %r9,%r2
    2fb8:       a7 f4 fe 60             j       2c78 <j9gc_createJavaLangString+0x388>
    2fbc:       b9 02 00 11             ltgr    %r1,%r1
    2fc0:       e3 20 f1 07 00 90       llgc    %r2,263(%r15)
    2fc6:       a7 84 00 ea             je      319a <j9gc_createJavaLangString+0x8aa>
    2fca:       b9 02 00 22             ltgr    %r2,%r2
    2fce:       e3 10 70 00 00 04       lg      %r1,0(%r7)
    2fd4:       a7 84 00 ea             je      31a8 <j9gc_createJavaLangString+0x8b8>
    2fd8:       a7 29 00 00             lghi    %r2,0
    2fdc:       e3 20 f0 a0 00 24       stg     %r2,160(%r15)
    2fe2:       e3 40 f0 c0 00 04       lg      %r4,192(%r15)
    2fe8:       b9 04 00 3a             lgr     %r3,%r10
    2fec:       b9 04 00 69             lgr     %r6,%r9
    2ff0:       e3 50 f0 b0 00 04       lg      %r5,176(%r15)
    2ff6:       b9 04 00 2b             lgr     %r2,%r11
    2ffa:       e3 10 17 e8 00 04       lg      %r1,2024(%r1)
    3000:       0d e1                   basr    %r14,%r1
    3002:       a7 f4 fe 8c             j       2d1a <j9gc_createJavaLangString+0x42a>
    3006:       c2 2e 00 00 0a ff       clgfi   %r2,2815
    300c:       a7 c4 fe e1             jle     2dce <j9gc_createJavaLangString+0x4de>
    3010:       e3 10 1a 20 00 04       lg      %r1,2592(%r1)
    3016:       a7 28 00 01             lhi     %r2,1
    301a:       50 21 80 04             st      %r2,4(%r1,%r8)
    301e:       a7 f4 fe df             j       2ddc <j9gc_createJavaLangString+0x4ec>
    3022:       c2 2e 00 00 0a ff       clgfi   %r2,2815
    3028:       a7 24 00 d7             jh      31d6 <j9gc_createJavaLangString+0x8e6>
    302c:       e3 20 1a 00 00 04       lg      %r2,2560(%r1)
    3032:       58 10 f0 cc             l       %r1,204(%r15)
    3036:       a5 1a 80 00             oilh    %r1,32768
    303a:       50 12 80 04             st      %r1,4(%r2,%r8)
    303e:       e3 10 b0 08 00 04       lg      %r1,8(%r11)
    3044:       e3 20 13 20 02 04       lg      %r2,8992(%r1)
    304a:       a7 2f 00 01             cghi    %r2,1
    304e:       a7 84 00 1c             je      3086 <j9gc_createJavaLangString+0x796>
    3052:       e3 40 19 e8 00 04       lg      %r4,2536(%r1)
    3058:       e3 30 19 e0 00 04       lg      %r3,2528(%r1)
    305e:       eb 44 00 08 00 0d       sllg    %r4,%r4,8
    3064:       a5 34 7f ff             nihh    %r3,32767
    3068:       e3 10 11 30 00 04       lg      %r1,304(%r1)
    306e:       b9 04 00 2b             lgr     %r2,%r11
    3072:       e3 30 40 48 00 08       ag      %r3,72(%r4)
    3078:       e3 10 12 08 00 04       lg      %r1,520(%r1)
    307e:       0d e1                   basr    %r14,%r1
    3080:       e3 10 b0 08 00 04       lg      %r1,8(%r11)
    3086:       e3 20 19 e8 00 04       lg      %r2,2536(%r1)
    308c:       eb 22 00 08 00 0d       sllg    %r2,%r2,8
    3092:       e3 10 19 e0 00 04       lg      %r1,2528(%r1)
    3098:       a5 14 7f ff             nihh    %r1,32767
    309c:       e3 20 20 48 00 04       lg      %r2,72(%r2)
    30a2:       e3 11 20 00 00 02       ltg     %r1,0(%r1,%r2)
    30a8:       a7 74 fe 9a             jne     2ddc <j9gc_createJavaLangString+0x4ec>
    30ac:       e3 10 70 00 00 04       lg      %r1,0(%r7)
    30b2:       a7 39 00 19             lghi    %r3,25
    30b6:       a7 49 00 01             lghi    %r4,1
    30ba:       b9 04 00 2b             lgr     %r2,%r11
    30be:       e3 10 11 08 00 04       lg      %r1,264(%r1)
    30c4:       0d e1                   basr    %r14,%r1
    30c6:       b9 02 00 32             ltgr    %r3,%r2
    30ca:       a7 84 fe 9e             je      2e06 <j9gc_createJavaLangString+0x516>
    30ce:       e3 40 f0 e0 00 04       lg      %r4,224(%r15)
    30d4:       b9 04 00 2b             lgr     %r2,%r11
    30d8:       c0 e5 00 00 00 00       brasl   %r14,30d8 <j9gc_createJavaLangString+0x7e8>
    30de:       b9 02 00 a2             ltgr    %r10,%r2
    30e2:       a7 84 fe 92             je      2e06 <j9gc_createJavaLangString+0x516>
    30e6:       e3 10 b0 08 00 04       lg      %r1,8(%r11)
    30ec:       e3 20 13 18 02 04       lg      %r2,8984(%r1)
    30f2:       a7 2b ff f9             aghi    %r2,-7
    30f6:       c2 2e 00 00 00 01       clgfi   %r2,1
    30fc:       a7 24 00 24             jh      3144 <j9gc_createJavaLangString+0x854>
    3100:       e3 90 19 e8 00 04       lg      %r9,2536(%r1)
    3106:       e3 20 f0 f8 00 04       lg      %r2,248(%r15)
    310c:       e3 40 19 e0 00 04       lg      %r4,2528(%r1)
    3112:       eb 99 00 08 00 0d       sllg    %r9,%r9,8
    3118:       a5 44 7f ff             nihh    %r4,32767
    311c:       e3 10 11 30 00 04       lg      %r1,304(%r1)
    3122:       b9 04 00 5a             lgr     %r5,%r10
    3126:       e3 30 20 30 00 04       lg      %r3,48(%r2)
    312c:       b9 04 00 2b             lgr     %r2,%r11
    3130:       e3 40 90 48 00 08       ag      %r4,72(%r9)
    3136:       e3 10 11 f8 00 04       lg      %r1,504(%r1)
    313c:       0d e1                   basr    %r14,%r1
    313e:       e3 10 b0 08 00 04       lg      %r1,8(%r11)
    3144:       e3 20 19 e8 00 04       lg      %r2,2536(%r1)
    314a:       eb 22 00 08 00 0d       sllg    %r2,%r2,8
    3150:       e3 10 19 e0 00 04       lg      %r1,2528(%r1)
    3156:       a5 14 7f ff             nihh    %r1,32767
    315a:       e3 20 20 48 00 04       lg      %r2,72(%r2)
    3160:       e3 a1 20 00 00 24       stg     %r10,0(%r1,%r2)
    3166:       e3 10 b0 08 00 04       lg      %r1,8(%r11)
    316c:       e3 20 13 10 02 04       lg      %r2,8976(%r1)
    3172:       a7 2f 00 02             cghi    %r2,2
    3176:       a7 84 fe 33             je      2ddc <j9gc_createJavaLangString+0x4ec>
    317a:       e3 10 11 30 00 04       lg      %r1,304(%r1)
    3180:       b9 04 00 4a             lgr     %r4,%r10
    3184:       e3 30 f0 f8 00 04       lg      %r3,248(%r15)
    318a:       b9 04 00 2b             lgr     %r2,%r11
    318e:       e3 10 10 30 00 04       lg      %r1,48(%r1)
    3194:       0d e1                   basr    %r14,%r1
    3196:       a7 f4 fe 23             j       2ddc <j9gc_createJavaLangString+0x4ec>
    319a:       b9 02 00 22             ltgr    %r2,%r2
    319e:       a7 74 00 85             jne     32a8 <j9gc_createJavaLangString+0x9b8>
    31a2:       e3 10 70 00 00 04       lg      %r1,0(%r7)
    31a8:       a7 29 00 00             lghi    %r2,0
    31ac:       e3 20 f0 a0 00 24       stg     %r2,160(%r15)
    31b2:       e3 40 f0 c0 00 04       lg      %r4,192(%r15)
    31b8:       b9 04 00 3a             lgr     %r3,%r10
    31bc:       b9 04 00 69             lgr     %r6,%r9
    31c0:       e3 50 f0 b0 00 04       lg      %r5,176(%r15)
    31c6:       b9 04 00 2b             lgr     %r2,%r11
    31ca:       e3 10 15 c0 00 04       lg      %r1,1472(%r1)
    31d0:       0d e1                   basr    %r14,%r1
    31d2:       a7 f4 fd a4             j       2d1a <j9gc_createJavaLangString+0x42a>
    31d6:       e3 10 1a 20 00 04       lg      %r1,2592(%r1)
    31dc:       a7 28 00 01             lhi     %r2,1
    31e0:       50 21 80 04             st      %r2,4(%r1,%r8)
    31e4:       a7 f4 ff 2d             j       303e <j9gc_createJavaLangString+0x74e>
    31e8:       e3 10 1a 20 00 04       lg      %r1,2592(%r1)
    31ee:       a7 28 00 00             lhi     %r2,0
    31f2:       50 21 80 04             st      %r2,4(%r1,%r8)
    31f6:       a7 f4 fd f3             j       2ddc <j9gc_createJavaLangString+0x4ec>
    31fa:       eb 23 f0 b0 00 04       lmg     %r2,%r3,176(%r15)
    3200:       e3 50 b0 08 00 04       lg      %r5,8(%r11)
    3206:       a7 1b 00 01             aghi    %r1,1
    320a:       a7 ab 00 02             aghi    %r10,2
    320e:       e3 20 5b e8 04 87       dlg     %r2,19432(%r5)
    3214:       e3 50 5b 40 04 04       lg      %r5,19264(%r5)
    321a:       eb 23 f0 b0 00 24       stmg    %r2,%r3,176(%r15)
    3220:       a7 3b 00 04             aghi    %r3,4
    3224:       e3 10 f0 c8 00 20       cg      %r1,200(%r15)
    322a:       eb 33 00 02 00 0d       sllg    %r3,%r3,2
    3230:       e3 20 f0 b0 00 04       lg      %r2,176(%r15)
    3236:       e3 33 90 00 00 16       llgf    %r3,0(%r3,%r9)
    323c:       eb 33 50 00 00 0d       sllg    %r3,%r3,0(%r5)
    3242:       42 43 20 00             stc     %r4,0(%r3,%r2)
    3246:       a7 74 fd 4b             jne     2cdc <j9gc_createJavaLangString+0x3ec>
    324a:       a7 f4 fd 65             j       2d14 <j9gc_createJavaLangString+0x424>
    324e:       18 42                   lr      %r4,%r2
    3250:       a7 4a 00 01             ahi     %r4,1
    3254:       e3 50 f0 e0 00 04       lg      %r5,224(%r15)
    325a:       b9 04 00 2b             lgr     %r2,%r11
    325e:       88 40 00 01             srl     %r4,1
    3262:       b9 17 00 44             llgtr   %r4,%r4
    3266:       c0 e5 00 00 00 00       brasl   %r14,3266 <j9gc_createJavaLangString+0x976>
    326c:       b9 04 00 92             lgr     %r9,%r2
    3270:       a7 f4 fd 04             j       2c78 <j9gc_createJavaLangString+0x388>
    3274:       b9 16 00 42             llgfr   %r4,%r2
    3278:       b9 04 00 2b             lgr     %r2,%r11
    327c:       c0 e5 00 00 00 00       brasl   %r14,327c <j9gc_createJavaLangString+0x98c>
    3282:       b9 04 00 92             lgr     %r9,%r2
    3286:       a7 f4 fc f9             j       2c78 <j9gc_createJavaLangString+0x388>
    328a:       a7 49 00 00             lghi    %r4,0
    328e:       e3 30 f0 c0 00 04       lg      %r3,192(%r15)
    3294:       b9 04 00 2a             lgr     %r2,%r10
    3298:       c0 e5 ff ff e6 b4       brasl   %r14,0 <getUnicodeLength>
    329e:       e3 20 f0 c8 00 24       stg     %r2,200(%r15)
    32a4:       a7 f4 fc b2             j       2c08 <j9gc_createJavaLangString+0x318>
    32a8:       e3 20 f0 b0 00 04       lg      %r2,176(%r15)
    32ae:       a7 21 00 01             tmll    %r2,1
    32b2:       a7 84 00 b7             je      3420 <j9gc_createJavaLangString+0xb30>
    32b6:       e3 30 f0 c0 00 02       ltg     %r3,192(%r15)
    32bc:       a7 84 fd 2f             je      2d1a <j9gc_createJavaLangString+0x42a>
    32c0:       b9 04 00 02             lgr     %r0,%r2
    32c4:       e3 41 a0 00 00 90       llgc    %r4,0(%r1,%r10)
    32ca:       b9 16 00 21             llgfr   %r2,%r1
    32ce:       e3 20 f0 b8 00 24       stg     %r2,184(%r15)
    32d4:       a7 29 00 00             lghi    %r2,0
    32d8:       e3 20 f0 b0 00 24       stg     %r2,176(%r15)
    32de:       eb 61 00 01 00 0d       sllg    %r6,%r1,1
    32e4:       a7 4f 00 2f             cghi    %r4,47
    32e8:       18 54                   lr      %r5,%r4
    32ea:       a7 74 00 04             jne     32f2 <j9gc_createJavaLangString+0xa02>
    32ee:       a7 58 00 2e             lhi     %r5,46
    32f2:       e3 20 90 04 00 12       lt      %r2,4(%r9)
    32f8:       b9 94 00 55             llcr    %r5,%r5
    32fc:       a7 84 00 56             je      33a8 <j9gc_createJavaLangString+0xab8>
    3300:       a7 1b 00 01             aghi    %r1,1
    3304:       40 56 90 08             sth     %r5,8(%r6,%r9)
    3308:       e3 10 f0 c0 00 20       cg      %r1,192(%r15)
    330e:       a7 74 ff db             jne     32c4 <j9gc_createJavaLangString+0x9d4>
    3312:       e3 00 f0 b0 00 24       stg     %r0,176(%r15)
    3318:       a7 01 00 20             tmll    %r0,32
    331c:       b9 04 00 20             lgr     %r2,%r0
    3320:       a7 84 fc fd             je      2d1a <j9gc_createJavaLangString+0x42a>
    3324:       e3 00 f0 c0 00 04       lg      %r0,192(%r15)
    332a:       b9 04 00 40             lgr     %r4,%r0
    332e:       a7 4b ff ff             aghi    %r4,-1
    3332:       a7 44 fc f4             jl      2d1a <j9gc_createJavaLangString+0x42a>
    3336:       e3 10 90 04 00 12       lt      %r1,4(%r9)
    333c:       a7 74 00 ca             jne     34d0 <j9gc_createJavaLangString+0xbe0>
    3340:       e3 10 b0 08 00 04       lg      %r1,8(%r11)
    3346:       b9 04 00 62             lgr     %r6,%r2
    334a:       e3 50 1b e8 04 04       lg      %r5,19432(%r1)
    3350:       e3 a0 1b 44 04 58       ly      %r10,19268(%r1)
    3356:       eb 55 00 01 00 0c       srlg    %r5,%r5,1
    335c:       b9 16 00 34             llgfr   %r3,%r4
    3360:       a7 29 00 00             lghi    %r2,0
    3364:       b9 87 00 25             dlgr    %r2,%r5
    3368:       b9 04 00 13             lgr     %r1,%r3
    336c:       a7 1b 00 04             aghi    %r1,4
    3370:       eb 22 00 01 00 0d       sllg    %r2,%r2,1
    3376:       eb 11 00 02 00 0d       sllg    %r1,%r1,2
    337c:       e3 11 90 00 00 16       llgf    %r1,0(%r1,%r9)
    3382:       eb 11 a0 00 00 0d       sllg    %r1,%r1,0(%r10)
    3388:       b9 08 00 12             agr     %r1,%r2
    338c:       d5 01 10 00 d0 00       clc     0(2,%r1),0(%r13)
    3392:       a7 84 00 bc             je      350a <j9gc_createJavaLangString+0xc1a>
    3396:       a7 4b ff ff             aghi    %r4,-1
    339a:       a7 07 ff e1             brctg   %r0,335c <j9gc_createJavaLangString+0xa6c>
    339e:       e3 60 f0 b0 00 24       stg     %r6,176(%r15)
    33a4:       a7 f4 fc bb             j       2d1a <j9gc_createJavaLangString+0x42a>
    33a8:       e3 60 b0 08 00 04       lg      %r6,8(%r11)
    33ae:       a7 1b 00 01             aghi    %r1,1
    33b2:       eb 23 f0 b0 00 04       lmg     %r2,%r3,176(%r15)
    33b8:       e3 40 6b e8 04 04       lg      %r4,19432(%r6)
    33be:       eb 44 00 01 00 0c       srlg    %r4,%r4,1
    33c4:       e3 60 6b 40 04 04       lg      %r6,19264(%r6)
    33ca:       b9 87 00 24             dlgr    %r2,%r4
    33ce:       eb 23 f0 b0 00 24       stmg    %r2,%r3,176(%r15)
    33d4:       a7 3b 00 04             aghi    %r3,4
    33d8:       e3 10 f0 c0 00 20       cg      %r1,192(%r15)
    33de:       eb 33 00 02 00 0d       sllg    %r3,%r3,2
    33e4:       e3 20 f0 b0 00 04       lg      %r2,176(%r15)
    33ea:       eb 22 00 01 00 0d       sllg    %r2,%r2,1
    33f0:       e3 33 90 00 00 16       llgf    %r3,0(%r3,%r9)
    33f6:       eb 33 60 00 00 0d       sllg    %r3,%r3,0(%r6)
    33fc:       40 52 30 00             sth     %r5,0(%r2,%r3)
    3400:       a7 74 ff 62             jne     32c4 <j9gc_createJavaLangString+0x9d4>
    3404:       a7 f4 ff 87             j       3312 <j9gc_createJavaLangString+0xa22>
    3408:       b9 04 00 13             lgr     %r1,%r3
    340c:       a7 28 00 00             lhi     %r2,0
    3410:       a7 f4 fa f7             j       29fe <j9gc_createJavaLangString+0x10e>
    3414:       a7 18 00 01             lhi     %r1,1
    3418:       42 10 f1 07             stc     %r1,263(%r15)
    341c:       a7 f4 fb f6             j       2c08 <j9gc_createJavaLangString+0x318>
    3420:       e3 20 f0 c0 00 02       ltg     %r2,192(%r15)
    3426:       a7 84 fc 7a             je      2d1a <j9gc_createJavaLangString+0x42a>
    342a:       e3 00 f0 b0 00 04       lg      %r0,176(%r15)
    3430:       b9 16 00 21             llgfr   %r2,%r1
    3434:       e3 20 f0 b8 00 24       stg     %r2,184(%r15)
    343a:       e3 20 90 04 00 12       lt      %r2,4(%r9)
    3440:       eb 41 00 01 00 0d       sllg    %r4,%r1,1
    3446:       a7 29 00 00             lghi    %r2,0
    344a:       e3 20 f0 b0 00 24       stg     %r2,176(%r15)
    3450:       e3 51 a0 00 00 90       llgc    %r5,0(%r1,%r10)
    3456:       a7 84 00 0d             je      3470 <j9gc_createJavaLangString+0xb80>
    345a:       a7 1b 00 01             aghi    %r1,1
    345e:       40 54 90 08             sth     %r5,8(%r4,%r9)
    3462:       e3 10 f0 c0 00 20       cg      %r1,192(%r15)
    3468:       a7 74 ff e4             jne     3430 <j9gc_createJavaLangString+0xb40>
    346c:       a7 f4 ff 53             j       3312 <j9gc_createJavaLangString+0xa22>
    3470:       e3 60 b0 08 00 04       lg      %r6,8(%r11)
    3476:       a7 1b 00 01             aghi    %r1,1
    347a:       eb 23 f0 b0 00 04       lmg     %r2,%r3,176(%r15)
    3480:       e3 40 6b e8 04 04       lg      %r4,19432(%r6)
    3486:       eb 44 00 01 00 0c       srlg    %r4,%r4,1
    348c:       e3 60 6b 40 04 04       lg      %r6,19264(%r6)
    3492:       b9 87 00 24             dlgr    %r2,%r4
    3496:       eb 23 f0 b0 00 24       stmg    %r2,%r3,176(%r15)
    349c:       a7 3b 00 04             aghi    %r3,4
    34a0:       e3 10 f0 c0 00 20       cg      %r1,192(%r15)
    34a6:       eb 33 00 02 00 0d       sllg    %r3,%r3,2
    34ac:       e3 20 f0 b0 00 04       lg      %r2,176(%r15)
    34b2:       eb 22 00 01 00 0d       sllg    %r2,%r2,1
    34b8:       e3 33 90 00 00 16       llgf    %r3,0(%r3,%r9)
    34be:       eb 33 60 00 00 0d       sllg    %r3,%r3,0(%r6)
    34c4:       40 52 30 00             sth     %r5,0(%r2,%r3)
    34c8:       a7 74 ff b4             jne     3430 <j9gc_createJavaLangString+0xb40>
    34cc:       a7 f4 ff 23             j       3312 <j9gc_createJavaLangString+0xa22>
    34d0:       b9 04 00 42             lgr     %r4,%r2
    34d4:       b9 04 00 20             lgr     %r2,%r0
    34d8:       a7 2b 00 03             aghi    %r2,3
    34dc:       b9 04 00 30             lgr     %r3,%r0
    34e0:       eb 22 00 01 00 0d       sllg    %r2,%r2,1
    34e6:       41 22 90 00             la      %r2,0(%r2,%r9)
    34ea:       d5 01 20 00 d0 00       clc     0(2,%r2),0(%r13)
    34f0:       b9 04 00 12             lgr     %r1,%r2
    34f4:       a7 84 00 14             je      351c <j9gc_createJavaLangString+0xc2c>
    34f8:       a7 2b ff fe             aghi    %r2,-2
    34fc:       a7 37 ff f7             brctg   %r3,34ea <j9gc_createJavaLangString+0xbfa>
    3500:       e3 40 f0 b0 00 24       stg     %r4,176(%r15)
    3506:       a7 f4 fc 0a             j       2d1a <j9gc_createJavaLangString+0x42a>
    350a:       e3 60 f0 b0 00 24       stg     %r6,176(%r15)
    3510:       a7 28 00 2f             lhi     %r2,47
    3514:       40 20 10 00             sth     %r2,0(%r1)
    3518:       a7 f4 fc 01             j       2d1a <j9gc_createJavaLangString+0x42a>
    351c:       e3 40 f0 b0 00 24       stg     %r4,176(%r15)
    3522:       a7 f4 ff f7             j       3510 <j9gc_createJavaLangString+0xc20>
    3526:       07 07                   nopr    %r7
```

The current inlined hash computation happens starting at `2e6e` where we can see the shift left by 5 and the subtraction, etc. in a tight loop. The branch at `2e6a` goes to the `else` path which lands at `29fe` which dispatches the call to `computeHashForUTF8` via the `BRASL` instruction at `2a3c`

Reply:
> Unfortunately at least on Linux on Z GCC is unable to inline past the function pointer table.

Right, which is why I suggested moving the implementation of computeHashForUTF8 into a header file (VMHelpers.hpp) so the code can be directly inlined.

Take the implementation from https://github.com/eclipse/openj9/blob/5b35623d452404eb3b5c8e245735af5ca935e602/runtime/util/utf8hash.c#L36 and move it to be close to https://github.com/eclipse/openj9/blob/5b35623d452404eb3b5c8e245735af5ca935e602/runtime/oti/VMHelpers.hpp#L975

We'll still need the c-callable version so we'll have to update the util/ implementation to call the one in the headerfile.  Which will be ugly as that's cross directory.  Might be worth checking if anyone currently uses the direct implementation from util/ as we can otherwise move it out.

Some code re-arranging will be required to make this work out nicely.

Reply:
I've tried to address this in 66bd2a52, however it seems there are several problems:

1. There is now duplication between VMHelpers.hpp and utfhash.c
2. It does not seem possible to call the VMHelpers version of the API from within utfhash because the former is a C++ file and the latter is a C file
3. There are many uses of `computeHashForUTF8` which indirect through the function table which we are unable to change due to include path issues. For example the uses in some of the files in [1] cannot be changed to use the VMHelpers version, even though they're C++ targets because VMHelpers.hpp relies on ObjectAllocationAPI.hpp being on the path, and it seems the files in shared_common do not include the gc_include directory on their IPATH.

I think 3. can be solved by adding gc_include to the IPATH of the targets in shared_common. I'm not experience in the IPATH area of the VM so I'm not sure if it's possible but shouldn't all the VM files have the same IPATHs so we don't run into such issues?

As for 1. and 2. I see duplication in other APIs such as `encodeUTF8Char`, `encodeUTF8CharN`, `decodeUTF8Char`, etc. which are all implemented in utf8*.c files as well. The duplication seems to be a much more general problem than seen here so it seems fixing such an issue should be outside the scope of this PR.

[1] https://github.com/eclipse/openj9/tree/master/runtime/shared_common